### PR TITLE
fix ffmpeg buffer seek in AudioFileReader.cpp

### PR DESCRIPTION
**Pull request by ngokevin**
The offset passed into BufferRead for my test WAV file large (for a really small WAV file) such that it seemed to seek to the end in one pass:
 
| Action | Old Pos | New Pos |
| SEEK_SET | 4096 | 3904414459037464
| SEEK_SET | 23334 | 3904418746140338
| SEEK_SET | 4096 | 3904418746163672
| SEEK_SET | 23334 | 3904423033266546
| SEEK_SET| 4096 | 3904423033289880

`offset` or `newPos` would be way larger than the buffer size. This line:

    newPos = std::min<int64_t>(std::max<int64_t>(newPos, 0), appData->data.size() - appData->dataPos);

Would evaluate to:

    min(max(someReallyLargeNumber, 0), size - currentDataPos);

Which would choose `size - currentDataPos`. Even though ffmpeg wanted to seek past the end, this `newPos` would always be less than the size of the data buffer, which told ffmpeg to keep seeking in an infinite cycle.

### CODE REVIEW

File: labsound/src/extended/AudioFileReader.cpp [hunk 1]
```diff
@@ -141,12 +141,17 @@ namespace detail
...
+    /**
+     * Tell ffmpeg how to seek through the buffer.
+     */
     int64_t AppData::bufferSeek(void *opaque, int64_t offset, int whence) {
       AppData *appData = (AppData *)opaque;
       if (whence == AVSEEK_SIZE) {
         return appData->data.size();
       } else {
         int64_t newPos;
+        int64_t oldPos = appData->dataPos;
```
Comment (ngokevin) [inline labsound/src/extended/AudioFileReader.cpp]: removing

File: labsound/src/extended/AudioFileReader.cpp [hunk 2]
```diff
@@ -156,7 +160,14 @@ namespace detail
         } else {
           newPos = offset;
         }
-        newPos = std::min<int64_t>(std::max<int64_t>(newPos, 0), appData->data.size() - appData->dataPos);
+
+        if (newPos > appData->data.size()) {
```
Comment (avaer) [inline labsound/src/extended/AudioFileReader.cpp]: Would it be more accurate to say that the previous min/max, now in the else, was simply wrong?

It was subtracting appData->dataPos from the min() bound but the real max is probably just the data size, as in this fix.

That is I think maybe we don’t need this if-else, but just fix the old max bound.

**Review by avaer — COMMENTED** [review]
This looks probably like a fix, just a question about the implementation.

File: labsound/src/extended/AudioFileReader.cpp (hunk 2 above)
Comment (ngokevin) [reply labsound/src/extended/AudioFileReader.cpp]: That’s actually what I did before but wasn’t sure why it was like that before, so I just applied it only if it was more than the max altogether. If you think it’s safe to assume, I can just set the min to the data size.

Comment (avaer) [reply labsound/src/extended/AudioFileReader.cpp]: Yeah I think the old code was just confusing the "new position" with "amount remaining".

I can't think of a reason for the old code being right.

Comment (ngokevin) [reply labsound/src/extended/AudioFileReader.cpp]: Updated.

**Review by avaer — APPROVED** [review]
Thanks for this, I'll rebuild + release.

**Pull request merged**